sql/analyzer/resolve_column_defaults.go

```diff
@@ -0,0 +1,488 @@
+	"st_union": {},
+	"st_validate": {},
+	"st_within": {},
+	"st_x": {},
+	"st_y": {},
+	"statement_digest": {},
+	"statement_digest_text": {},
+	"std": {},
+	"stddev": {},
+	"stddev_pop": {},
+	"stddev_samp": {},
+	"str_to_date": {},
+	"strcmp": {},
+	"subdate": {},
+	"substr": {},
+	"substring": {},
+	"substring_index": {},
+	"subtime": {},
+	"sum": {},
+	"sysdate": {},
+	"system_user": {},
+	"tan": {},
+	"time": {},
+	"time_format": {},
+	"time_to_sec": {},
+	"timediff": {},
+	"timestamp": {},
+	"timestampadd": {},
+	"timestampdiff": {},
+	"to_base64": {},
+	"to_days": {},
+	"to_seconds": {},
+	"trim": {},
+	"truncate": {},
+	"ucase": {},
+	"uncompress": {},
+	"uncompressed_length": {},
+	"unhex": {},
+	"unix_timestamp": {},
+	"updatexml": {},
+	"upper": {},
+	"user": {},
+	"utc_date": {},
+	"utc_time": {},
+	"utc_timestamp": {},
+	"uuid": {},
+	"uuid_short": {},
+	"uuid_to_bin": {},
+	"validate_password_strength": {},
+	"values": {},
+	"var_pop": {},
+	"var_samp": {},
+	"variance": {},
+	"version": {},
+	"wait_for_executed_gtid_set": {},
+	"wait_until_sql_thread_after_gtids": {},
+	"week": {},
+	"weekday": {},
+	"weekofyear": {},
+	"weight_string": {},
+	"year": {},
+	"yearweek": {},
+}
+
+func resolveColumnDefaults(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+	span, _ := ctx.Span("resolveColumnDefaults")
+	defer span.Finish()
+
+	return plan.TransformUp(n, func(n sql.Node) (sql.Node, error) {
+		switch node := n.(type) {
+		case sql.SchemeModifiable:
+			sch := node.Schema()
+			var columns = make(map[string]indexedCol)
+			for i, col := range sch {
+				columns[strings.ToLower(col.Name)] = indexedCol{col, i}
+			}
+			newSch := make(sql.Schema, len(sch))
+
+			for i, col := range sch {
+				newCol := *col
+				if col.Default == nil || col.Default.Resolved() {
+					newSch[i] = &newCol
+					continue
+				}
+				if sql.IsTextBlob(newCol.Type) && newCol.Default.IsLiteral() {
+					return nil, fmt.Errorf("text/blob types may only have expression default values")
+				}
+				newDefaultExpr, err := expression.TransformUp(newCol.Default.Expression, func(e sql.Expression) (sql.Expression, error) {
+					switch expr := e.(type) {
+					case *expression.UnresolvedFunction:
+						funcName := expr.Name()
+						if _, isValid := validColumnDefaultFuncs[funcName]; !isValid {
+							return nil, fmt.Errorf("function `%s` on column `%s` is not valid for usage in a default value", funcName, col.Name)
+						}
+						if (funcName == "now" || funcName == "current_timestamp") &&
+							newCol.Default.IsLiteral() &&
+							(!sql.IsTime(newCol.Type) || sql.Date == newCol.Type) {
+							return nil, fmt.Errorf("only datetime/timestamp may declare default values of now()/current_timestamp() without surrounding parentheses")
+						}
+						validFunc, err := a.Catalog.Function(funcName)
+						if err != nil {
+							return nil, err
+						}
+						return validFunc.Call(expr.Arguments...)
+					case *expression.UnresolvedColumn:
```

Comment:
Same comment. This one is actually more important, because the schema of the underlying nodes might actually get pruned in some strange cases (columns moved or deleted). Affected nodes really need to implement sql.Expressioner to avoid getting caught like this.

So when the analyzer gets to this stage, all columns should be resolved, and this should just perform validation. (like you can only refer to columns in the same table / same type)